### ISSUE TITLE
Can't run "Fuzzing with AFL" example (MacOS)

### ISSUE THREAD

**Issue by padolph** [rohanpadhye/jqf#39]
Following the tutorial on https://github.com/rohanpadhye/jqf/wiki/Fuzzing-with-AFL on macOS 10.14.5 (Mojave), I was unable to get the example to run successfully.

I'm not sure if this is a problem with jqf or my install of afl. The jqf -v option just created an empty jqf.log file so no insight there. The jqf -t option did not help.

```
$ jqf/bin/jqf-afl-fuzz -i afl/testcases/images/png/ -c $(jqf/scripts/classpath.sh):<classpath> com.foo.PngTest testRead
Performing pilot run....  Pilot run success! Launching AFL now...
afl-fuzz 2.52b by <lcamtuf@google.com>
[+] You have 4 CPU cores and 3 runnable tasks (utilization: 75%).
[+] Try parallel jobs - see docs/parallel_fuzzing.txt.
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Scanning 'afl/testcases/images/png/'...
[+] No auto-generated dictionary tokens to reuse.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Attempting dry run with 'id:000000,orig:not_kitty.png'...
[!] WARNING: Test case results in a timeout (skipping)
[*] Attempting dry run with 'id:000001,orig:not_kitty_alpha.png'...
[!] WARNING: Test case results in a timeout (skipping)
[*] Attempting dry run with 'id:000002,orig:not_kitty_gamma.png'...
[!] WARNING: Test case results in a timeout (skipping)
[*] Attempting dry run with 'id:000003,orig:not_kitty_icc.png'...
[!] WARNING: Test case results in a timeout (skipping)

[-] PROGRAM ABORT : All test cases time out, giving up!
         Location : perform_dry_run(), afl-fuzz.c:2883
```

**Comment by rohanpadhye** [comment]
Hmm. That's strange. I just tried running this again on MacOS and it works fine for me without `-t`. Can you tell me the output of `java -version`? And could you also confirm if you are using a recent version of AFL? 
![image](https://user-images.githubusercontent.com/4233266/59731654-e5384600-91fb-11e9-8e0f-bbc12237f05a.png)

PS: How long does it take to report the timeout for you in each test case? Is it proportional to the amount you give it in `-t` or does it report timeout instantly?

**Comment by padolph** [comment]
Thanks for your reply.

$ java -version
java version "1.8.0_172"
Java(TM) SE Runtime Environment (build 1.8.0_172-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.172-b11, mixed mode)

$ ./afl/afl-clang++ --version
afl-cc 2.52b by <lcamtuf@google.com>
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

The run takes no time at all so I'm sure it is not a real timeout. It
reports the timeouts instantly, and -t has no effect on this. Also note
with -v the log file remains empty, so it must be failing really early. I
tried also with my own java class and had the same results.

There might be a problem with my classpath though - I will check this and
report tomorrow. Still, I would not expect to get beyond the pilot run if
that was the case.

Thank and best regards,

-Paul

**Comment by padolph** [comment]
I did have a problem with my classpath but fixed that. I confirmed that I
can run both the tutorial Png test driver and my own test driver on the
command line with java -cp with no issues. Using this same classpath plus
the extra entries from jqf/scripts/classpath.sh, however, I am still seeing
the fail instantly problem with jqf.

Any more clues that I can search for?

**Comment by rohanpadhye** [comment]
Thanks for the version info. That looks fine. I've pushed another option for debugging the AFL/JVM proxy. Could you please `git pull` in the JQF repo and then re-run the `jqf-afl-fuzz` command but with an additional `-a` option? This option should create a file `proxy.log` that logs the inter-process communication between AFL and JQF. The top of my `proxy.log` for the tutorial looks like this:

```
opened to java fifo /tmp/jqf.Xex/a2j
opened from java fifo /tmp/jqf.Xex/j2a
Said hello to AFL (init).
Said hello to AFL (in loop).
Said hello to Java.
Got return status from Java.
Got trace bits from java.
sent return status to AFL.
Said hello to AFL (in loop).
Said hello to Java.
Got return status from Java.
Got trace bits from java.
sent return status to AFL.
Said hello to AFL (in loop).
.... (keeps going on; ~6 lines per test execution) ....
```

Thanks.

**Comment by padolph** [comment]
I pulled from master and re-ran jqf/setup.sh as directed by the tutorial.
The build was successful and my git log looks like this:

2bfcde4 2019-06-19 (HEAD -> master, origin/master, origin/HEAD) README:
Update paper info [Rohan Padhye]
2b85061 2019-06-19 jqf-afl-fuzz: Support -a option for afl-proxy debugging
[Rohan Padhye]
ad18de7 2019-06-18 (origin/zeal) ExecutionIndexing: Identify minimized
inputs [Rohan Padhye]
32f273b 2019-06-18 Coverage: Implement hashCode() and nonZeroHashCode()
[Rohan Padhye]

So I'm pretty sure I have your changes. I also see the new -a option in the
help string. However, using the -a flag I am not getting any proxy.log file
created.

Also recall that with -v I get only an empty jqf.log. Could this be some
permissions issue on my Mac that is not allowing jqf to interact with afl
or even start it?

Thanks.

**Comment by rohanpadhye** [comment]
The AFL proxy basically mimics AFL's forkserver. In your original post, I notice that AFL did not print the message `[*] Spinning up the fork server...`.  Ideally, it should say this after `[*] Validating target binary...` and before `[*] Attempting dry run with [...]`. Have you disabled AFL fork server explicitly? I know that AFL lets you do this via an environment variable.

To test this possibility, I set `export AFL_NO_FORKSRV=1` on my machine and was able to reproduce the timeouts that you saw! When I deleted this env var, the timeouts went away and the tutorial works fine. If you have any env vars `AFL_*` set, please remove them before trying JQF.

**Comment by padolph** [comment]
That was it! Thanks so much!!

I had set that environment variable because a somewhat forceful AFL warning
message told me to do so, citing poor performance of the macOS fork
implementation. I interpreted this as advice to speed up execution not kill
it! I shouldn't have blindly followed it...

JQF is running now on my custom test driver. AFL still warns me about the
"target is very slow" or somesuch, probably related again to the macOS fork
issue. If speed turns out to be a problem I will move to a linux VM.

But I am happily fuzzing away now.

Thanks again.
